src/dbnode/storage/index/aggregate_results.go

```diff
@@ -101,201 +95,114 @@ func (r *aggregatedResults) Reset(
+		if remaining := r.aggregateOpts.SizeLimit - r.size; remaining < maxInserts {
+			maxInserts = remaining
+		}
+	}
+
+	limitTripped := false
+	docs := 0
+	numInserts := 0
 	for _, entry := range batch {
+		if docs >= maxDocs || numInserts >= maxInserts {
+			limitTripped = true
+		}
+
+		if limitTripped {
+			entry.Field.Finalize()
+			for _, term := range entry.Terms {
+				term.Finalize()
+			}
+
+			r.size += numInserts
+			r.totalDocsCount += docs
+			return r.size, r.totalDocsCount
+		}
+
+		docs++
 		f := entry.Field
 		aggValues, ok := r.resultsMap.Get(f)
 		if !ok {
-			aggValues = r.valuesPool.Get()
-			// we can avoid the copy because we assume ownership of the passed ident.ID,
-			// but still need to finalize it.
-			r.resultsMap.SetUnsafe(f, aggValues, AggregateResultsMapSetUnsafeOptions{
-				NoCopyKey:     true,
-				NoFinalizeKey: false,
-			})
-		} else {
-			// because we already have a entry for this field, we release the ident back to
-			// the underlying pool.
-			f.Finalize()
-		}
-		valuesMap := aggValues.Map()
-		for _, t := range entry.Terms {
-			if !valuesMap.Contains(t) {
+			if maxInserts > numInserts {
+				numInserts++
+				aggValues = r.valuesPool.Get()
 				// we can avoid the copy because we assume ownership of the passed ident.ID,
 				// but still need to finalize it.
-				valuesMap.SetUnsafe(t, struct{}{}, AggregateValuesMapSetUnsafeOptions{
+				r.resultsMap.SetUnsafe(f, aggValues, AggregateResultsMapSetUnsafeOptions{
 					NoCopyKey:     true,
 					NoFinalizeKey: false,
 				})
-				valueInsertions++
 			} else {
-				// because we already have a entry for this term, we release the ident back to
-				// the underlying pool.
-				t.Finalize()
+				// this value exceeds the limit, so should be released to the underling
+				// pool without adding to the map.
+				f.Finalize()
 			}
+		} else {
+			// because we already have a entry for this field, we release the ident back to
+			// the underlying pool.
+			f.Finalize()
 		}
-	}
-	size := r.resultsMap.Len()
-	docsCount := r.totalDocsCount + valueInsertions
-	r.totalDocsCount = docsCount
-	r.Unlock()
-	return size, docsCount
-}
 
-func (r *aggregatedResults) addDocumentsBatchWithLock(
-	batch []doc.Document,
-) error {
-	for _, doc := range batch {
-		switch r.aggregateOpts.Type {
-		case AggregateTagNamesAndValues:
-			if err := r.addDocumentWithLock(doc); err != nil {
-				return err
-			}
-
-		case AggregateTagNames:
-			// NB: if aggregating by name only, can ignore any additional documents
-			// after the result map size exceeds the optional size limit, since all
-			// incoming terms are either duplicates or new values which will exceed
-			// the limit.
-			size := r.resultsMap.Len()
-			if r.aggregateOpts.SizeLimit > 0 && size >= r.aggregateOpts.SizeLimit {
-				return nil
-			}
-
-			if err := r.addDocumentTermsWithLock(doc); err != nil {
-				return err
+		valuesMap := aggValues.Map()
+		for _, t := range entry.Terms {
+			if maxDocs > docs {
```

Comment:
if the max is already hit should we break from the `for`? Or is it a trivial amount of iteration?

Reply:
We still have to iterate through the remainder here to finalize them